tools/ci-build/publisher/src/package.rs

```diff
@@ -15,22 +15,90 @@ use smithy_rs_tool_common::package::PackageCategory;
 use std::collections::{BTreeMap, BTreeSet, HashSet};
 use std::error::Error as StdError;
 use std::fmt;
+use std::fmt::Formatter;
 use std::path::{Path, PathBuf};
+use std::str::FromStr;
 use tokio::fs;
 use tracing::warn;
 
+/// A trait that marks a type that can be created from `&str` and turned into `LenientVersion`
+pub trait ParseIntoLenientVersion {
+    type ParsedT: FromStr<Err = Self::Err> + Into<LenientVersion>;
+    type Err: Into<BoxError>;
+}
+
+/// A type that indicates the `SemVer` variant of `LenientVersion` will be created as a result of
+/// parsing
+pub struct SemVer;
+impl ParseIntoLenientVersion for SemVer {
+    type ParsedT = Version;
+    type Err = semver::Error;
+}
+
+/// A type that indicates the `Lenient` variant of `LenientVersion` will be created as a result of
+/// parsing
+pub struct Lenient;
+impl ParseIntoLenientVersion for Lenient {
+    type ParsedT = String;
+    type Err = std::convert::Infallible;
```

Comment:
is this trait really required? I'm trying to understand how it's actually used...

Reply:
It is more of a convenience trait, and without it, call site like below
```
read_package<SemVer>(...)
```
would need to specify
```
read_package<semver::Version, semver::Error>(...)
```
and functions being called down the road need to have two generic parameters instead of one.

I'll add a comment as such.

Reply:
Added a comment in 0a4ab75.